### PR TITLE
fix(catch): fix catch to dispose old subscriptions 

### PR DESCRIPTION
**Pull request by staltz**
**DEPENDS ON PR #764 MERGED FIRST.**

Fix catch operator to not have anymore a shared underlying Subscription, and instead reset the
subscription for each new observable replacing the caught error. This fixes a potential memory leak
if catch is used as an infinite retry, because subscriptions would be retained since the beginning,
and would increasing each time a catch is performed.

Resolves issue #763.

### CODE REVIEW

File: src/operators/catch.ts [hunk 1]
```diff
@@ -50,7 +50,17 @@ class CatchSubscriber<T> extends Subscriber<T> {
...
+      this.lastSubscription = result.subscribe(this.destination);
     }
   }
+
+  _complete() {
+    this.lastSubscription.unsubscribe();
+    this.destination.complete();
+  }
+
+  _unsubscribe() {
```
Comment (kwonoj) [inline src/operators/catch.ts]: Is this function being called?

Comment (staltz) [inline src/operators/catch.ts]: According to test coverage, yes it is. `_unsubscribe()` is overriding Subscription._unsubscribe(), which is called inside Subscription.unsubscribe().

Comment (kwonoj) [inline src/operators/catch.ts]: Ah, didn't aware it's inherited from `Subscription`. thought it's inherited from `Subscriber` and tried to look `Subscriber::_unsubscribe()` instead.. my bad. :(

**Comment by kwonoj** [comment]
@staltz , would you able to rebase this one? #764 is now merged.

**Comment by staltz** [comment]
Rebased

**Comment by kwonoj** [comment]
Merged with 280f7ed, thanks @staltz

**Pull request closed without merging**